core/src/test/scala/neotypes/AsyncGuaranteeSpec.scala

```diff
@@ -0,0 +1,124 @@
+package neotypes
+
+import neotypes.internal.syntax.async._
+
+import org.scalatest.{EitherValues, FlatSpec, Matchers}
+import org.scalatest.compatible.Assertion
+
+/** Base class for testing the Async[F].guarantee method. */
+abstract class AsyncGuaranteeSpec[F[_]] extends FlatSpec with Matchers with EitherValues {
+  def fToEither[T](f: F[T]): Either[Throwable, T]
+  implicit def F: Async[F]
+
+  trait AsyncGuaranteeFixture {
+    private[this] var counter = 0
+
+    final def assertFinalizerWasCalledOnlyOnce: Assertion =
+      withClue("Finalizer was called more than once -") {
+        this.counter shouldBe 1
+      }
+
+    final def assertFinalizerWasNotCalled: Assertion =
+      withClue("Finalizer was called -") {
+        this.counter shouldBe 0
+      }
+
+    final def run[T](result: Either[Throwable, T],
+                     inputEx: Option[Throwable] = None,
+                     finalizerEx: Option[Throwable] = None): Either[Throwable, T] = {
+      def fromOption(opt: Option[Throwable]): F[Unit] =
+        opt.fold(ifEmpty = F.delay(()))(ex => F.failed(ex))
+
+      fToEither(
+        F.guarantee[Unit, T](
+          fa = fromOption(inputEx)
+        )(
+          _ => result.fold(ex => F.failed(ex), t => F.delay(t))
+        ) { (_, _) =>
+          F.delay {
+            counter += 1
+          } flatMap { _ =>
+            fromOption(finalizerEx)
+          }
+        }
+      )
+    }
+  }
+
+  it should "execute finalizer and return the result when nothing fails" in new AsyncGuaranteeFixture {
+    val expectedResult = "result"
+
+    val res = run(result = Right(expectedResult))
+
+    res.right.value shouldBe expectedResult
+    assertFinalizerWasCalledOnlyOnce
+  }
+
+  it should "not execute finalizer and return input exception when input fails" in new AsyncGuaranteeFixture {
+    val inputEx = new Exception("Input failed")
+
+    val ex = run(inputEx = Some(inputEx), result = Right("result"))
+
```

Comment:
the run method is defined with parameter "result" first and "inputEx" second.  It's just stylistic, but I would keep the order the same as the method definition, so val ex = run( result = Right("result"), inputEx = Some(inputEx))

Reply:
@geoffjohn11 Yeah, I thought about that, and I end up making the changes like this. Because I like the idea of `input` -> `use` -> `close`. In the original method, it has to have the other order because `result` doesn't have a default value, so it has to go firts.